### ISSUE TITLE
Shared axes removing label information from `self.ax`

### ISSUE THREAD

**Issue by kevindougherty-noaa** [NOAA-EMC/emcpy#76]
I have been doing a bit of testing and ran into an issue when trying to plot two lines on a shared y-axis. 
```
x1 = [1, 2, 3, 4]
y1 = [1, 2, 3, 4]

x2 = [10, 15, 20, 25]
y2 = [0, 5, 10, 15]

# Create line plot 1 object
lp1 = LinePlot(x1, y1)
lp1.label = 'line 1'

# Create line plot 2 object
lp2 = LinePlot(x2, y2)
lp2.color = 'tab:green'
lp2.label = 'line 2'
lp2.use_shared_ay() # add shared y axis

# Create plot and draw data
myplt = CreatePlot()
plt_list = [lp1, lp2]
myplt.draw_data(plt_list)

# Add features
myplt.add_title(label='Test Line Plot, 2 X Axes ')
myplt.add_xlabel(xlabel='X Axis Label')
myplt.add_ylabel(ylabel='Y Axis Label')
myplt.add_legend(loc='upper left')

# Add secondary xaxis
myplt.add_xlabel(xlabel='Secondary X Axis Label', xaxis='secondary')
```
![image](https://user-images.githubusercontent.com/69815622/138506503-daa4054d-80ab-436f-b5d3-f27071f66f70.png)

You can see the legend does not include the second line's label. When I do a bit more digging, I find that that the text is not included when checking:
```
myplt.ax.legend_.__dict__
```
```
{'_stale': False,
 'stale_callback': <function matplotlib.artist._stale_axes_callback(self, val)>,
 '_axes': <AxesSubplot:title={'center':'Test Line Plot, 2 X Axes '}, xlabel='X Axis Label', ylabel='Y Axis Label'>,
 'figure': <Figure size 576x432 with 2 Axes>,
 '_transform': <matplotlib.transforms.IdentityTransform at 0x7fc8d0907450>,
 '_transformSet': False,
 '_visible': True,
 '_animated': False,
 '_alpha': None,
 'clipbox': None,
 '_clippath': None,
 '_clipon': True,
 '_label': '',
 '_picker': None,
 '_contains': None,
 '_rasterized': None,
 '_agg_filter': None,
 '_mouseover': False,
 'eventson': False,
 '_oid': 0,
 '_propobservers': {},
 '_remove_method': <bound method Axes._remove_legend of <AxesSubplot:title={'center':'Test Line Plot, 2 X Axes '}, xlabel='X Axis Label', ylabel='Y Axis Label'>>,
 '_url': None,
 '_gid': None,
 '_snap': None,
 '_sketch': None,
 '_path_effects': [],
 '_sticky_edges': _XYPair(x=[], y=[]),
 '_in_layout': True,
 'prop': <matplotlib.font_manager.FontProperties at 0x7fc8d0a9d190>,
 '_fontsize': 10.0,
 'texts': [Text(0, 0, 'line 1')],
 'legendHandles': [<matplotlib.lines.Line2D at 0x7fc8d091a990>],
 '_legend_title_box': <matplotlib.offsetbox.TextArea at 0x7fc8d091a6d0>,
 '_custom_handler_map': None,
 'numpoints': 1,
 'markerscale': 1.0,
 'shadow': False,
 'columnspacing': 2.0,
 'scatterpoints': 1,
 'handleheight': 0.7,
 'borderpad': 0.4,
 'labelspacing': 0.5,
 'handlelength': 2.0,
 'handletextpad': 0.8,
 'borderaxespad': 0.5,
 '_ncol': 1,
 '_scatteryoffsets': array([0.375]),
 '_legend_box': <matplotlib.offsetbox.VPacker at 0x7fc8d091a4d0>,
 'isaxes': True,
 'parent': <AxesSubplot:title={'center':'Test Line Plot, 2 X Axes '}, xlabel='X Axis Label', ylabel='Y Axis Label'>,
 '_loc_used_default': False,
 '_mode': None,
 '_bbox_to_anchor': None,
 'legendPatch': <matplotlib.patches.FancyBboxPatch at 0x7fc8d0718090>,
 '_legend_handle_box': <matplotlib.offsetbox.HPacker at 0x7fc8d06efe90>,
 '_loc_real': 2,
 '_draggable': None}
```

It can be seen that `'texts': [Text(0, 0, 'line 1')],` only includes the first object's label.

**Comment by kevindougherty-noaa** [comment]
@EdwardSafford-NOAA When you get a chance, would you be able to take a look at this?

**Comment by EdwardSafford-NOAA** [comment]
@kevindougherty-noaa If there's no rush then yes I can.  It will probably be a week or two before I can get to it.

It's a known problem from #63.  A less than optimal solution is to plot the legend from the figure rather than the plot, which is possible as of Matplotlib 2.1 (below).   It's also possible to save off the legends for every plotted figure on all axes and then send those into an add_legend call, which would yield a more standard legend.

![image](https://user-images.githubusercontent.com/62339196/138509267-9c8cfdf7-e89b-475e-b5ee-520bb2698d49.png)

**Comment by EdwardSafford-NOAA** [comment]
I have two ideas for a solution.  The cleanest way to handle this is to create the legend from the figure instead of from a specific axis (subplot) object.  Here I've plotted 3 lines with lines 2 and 3 using a shared y axis and added the standard locations for the legend (i.e. left center, center, right center, etc):
![test_line_plot_2_x_axes](https://user-images.githubusercontent.com/62339196/140105316-c44cb4da-b319-45c1-8910-005c301f0dd1.png)

That can be improved by adding a bbox and transforming the coordinates to the default drawing axes:
![test_line_plot_2_x_axes(2)](https://user-images.githubusercontent.com/62339196/140105541-9cc5ee4d-8ba5-4f41-8426-a67e4b55f38f.png)


The only problem with this approach is that 'best' or 0 is not an available location for the legend.  Because it's getting added at the figure level it doesn't know where the lines are so it can't calculate 'best'.

There is another approach.  For every line that's plotted to a shared x or y axis an empty plot can be added to the default axes.  It has no coordinates to plot but does have the color and label settings.  So when ax.add_legend() will include all the labels:
![test_line_plot_2_x_axes(1)](https://user-images.githubusercontent.com/62339196/140109690-01de555f-9ed4-42f4-a9b2-7712b556d098.png)

The second approach seems a bit of a hack to me.  It's also worth noting that the default 'best' location would be applied using only those lines plotted to the default axes and not the shared_x and shared_y.  

I have one other idea I'm going to chase down before any decisions need to be made.

**Comment by kevindougherty-noaa** [comment]
@EdwardSafford-NOAA Thanks for tackling this. I think the first approach you used would be more than sufficient and it should be no problem that we don't have 'best'. I think it would be better to be more declarative in the location anyway and make the default 'upper left' or 'upper right'.

**Comment by EdwardSafford-NOAA** [comment]
@kevindougherty-noaa Ok I'll go that way then.  I'll extend the Figure class to add a add_legend() method which will handle the translations of the loc options to the axes.  Question:  should I keep the existing add_legend() method in the emcpy.plots.create_plots.CreatePlot class?  I can make an argument either way for retaining/removing it.

**Comment by kevindougherty-noaa** [comment]
@EdwardSafford-NOAA Where do the changes that fix this issue go now?

**Comment by EdwardSafford-NOAA** [comment]
@kevindougherty-noaa Creating a legend in a figure class object can be done directly, as I did to generate the prototype, but it's a bit cumbersome and a bit too much to expect of a user.  For example this code produces a legend in the upper right and lower center respectively:
`fig.legend(loc="upper left", bbox_to_anchor=(0,1), bbox_transform=myplt.ax.transAxes)`
`fig.legend(loc="lower center", bbox_to_anchor=(0.5,0), bbox_transform=myplt.ax.transAxes)`

Note fig is a figure object, myplt is a CreatePlot object.

My thought is to add a CreateFigure class to create_plots.py which will extend the matplotlib.figure class by adding an add_legend() method.  This would map a location string to the correct x,y location within the CreatePlot object so the user won't have to figure that part out.

**Comment by EdwardSafford-NOAA** [comment]
I think it's together now.  This is an example of the available locations using the new CreateFigure class, an extension of matplotlib.figure.Figure and it's add_legend() method.
![test_line_plot_2_x_axes(3)](https://user-images.githubusercontent.com/62339196/140795128-771c0772-8710-4092-abf0-1bbfa7ca6ec9.png)

I kept the add_legend() method in the CreatePlot class though I updated the test and example scripts to use the add_legend() method from the CreateFigure class.
